viogpu/viogpu3d/viogpu_command.cpp

```diff
@@ -0,0 +1,395 @@
+#include "viogpu_command.h"
+#include "viogpu_device.h"
+#include "viogpu_adapter.h"
+#include "baseobj.h"
+
+#pragma code_seg(push)
+#pragma code_seg()
+
+VioGpuCommand::VioGpuCommand(VioGpuAdapter* adapter) {
+    DbgPrint(TRACE_LEVEL_VERBOSE, ("<---> %s", __FUNCTION__));
+
+    m_pAdapter = adapter;
+    m_pCommander = &adapter->commander;
+    m_pContext = NULL;
+
+    m_FenceId = 0;
+    m_pDmaBuffer = NULL;
+    m_pCommand = NULL;
+    m_pEnd = NULL;
+
+    m_allocations = NULL;
+    m_allocationsLength = 0;
+
+    list_entry.Blink = NULL;
+    list_entry.Flink = NULL;
+};
+
+
+void VioGpuCommand::PrepareSubmit(const DXGKARG_SUBMITCOMMAND* pSubmitCommand) {
+    DbgPrint(TRACE_LEVEL_VERBOSE, ("<---> %s", __FUNCTION__));
+
+    m_FenceId = pSubmitCommand->SubmissionFenceId;
+    if (m_pDmaBuffer) {
+        m_pCommand = (char*)m_pDmaBuffer + pSubmitCommand->DmaBufferSubmissionStartOffset;
+        m_pEnd = (char*)m_pDmaBuffer + pSubmitCommand->DmaBufferSubmissionEndOffset;
+    }
+    m_pContext = reinterpret_cast<VioGpuDevice*>(pSubmitCommand->hContext);
+}
+
+#pragma code_seg(pop)
+
+PAGED_CODE_SEG_BEGIN
+
+void VioGpuCommand::Run() {
+    PAGED_CODE();
+    DbgPrint(TRACE_LEVEL_VERBOSE, ("<---> %s", __FUNCTION__));
+
+    while (m_pCommand < m_pEnd) {
+
+        VIOGPU_COMMAND_HDR* cmdHdr = (VIOGPU_COMMAND_HDR*)m_pCommand;
+        m_pCommand += sizeof(VIOGPU_COMMAND_HDR);
+
+        void* cmdBody = m_pCommand;
+        m_pCommand += cmdHdr->size;
+
+        DbgPrint(TRACE_LEVEL_VERBOSE, ("%s fence_id=%d running command=%d", __FUNCTION__, m_FenceId, cmdHdr->type));
+
+        switch (cmdHdr->type) {
+        case VIOGPU_CMD_SUBMIT: {
+            m_pAdapter->ctrlQueue.SubmitCommand(
+                cmdBody, cmdHdr->size, m_pContext->GetId(),
+                VioGpuCommand::QueueRunningCb, this);
```

Comment:
```suggestion
            PBYTE submitCmd = new(NonPagedPoolNx)BYTE[cmdHdr->size];
            RtlCopyMemory(submitCmd, cmdBody, cmdHdr->size);

            m_pAdapter->ctrlQueue.SubmitCommand(
                submitCmd, cmdHdr->size, m_pContext->GetId(),
                VioGpuCommand::QueueRunningCb, this);
```

Reply:
Fixed, it seems to have fix BSODs during driver unloads. This part (command submission) will have to be reworked anyway when driver will be ported to WDDMv2, so small small performance loss is not critical.